examples/advanced_operations/add_gmail_ad.rb

```diff
@@ -0,0 +1,182 @@
+  content_type.upcase.gsub("/", "_").to_sym
+end
+
+def add_gmail_ad(customer_id, campaign_id, ad_group_id)
+  client = Google::Ads::GoogleAds::GoogleAdsClient.new
+
+  logo_image_bytes, logo_image_content_type = get_image(
+    'https://goo.gl/mtt54n'
+  )
+  marketing_image_bytes, marketing_image_content_type = get_image(
+    'http://goo.gl/3b9Wfh'
+  )
+
+  logo_image_content_type = content_type_to_symbol(logo_image_content_type)
+  marketing_image_content_type = content_type_to_symbol(
+    marketing_image_content_type
+  )
+
+  media_file_service = client.service(:MediaFile)
+
+  media_file_logo = client.resource(:MediaFile)
+  media_file_logo.type = :IMAGE
+  media_file_logo.image = client.resource(:MediaImage)
+  media_file_logo.image.data = client.wrapper.bytes(logo_image_bytes)
+  media_file_logo.mime_type = logo_image_content_type.to_sym
+
+  media_file_logo_op = client.operation(:MediaFile)
+  media_file_logo_op['create'] = media_file_logo
+
+  media_file_marketing = client.resource(:MediaFile)
+  media_file_marketing.type = :IMAGE
+  media_file_marketing.image = client.resource(:MediaImage)
+  media_file_marketing.image.data = client.wrapper.bytes(marketing_image_bytes)
+  media_file_marketing.mime_type = marketing_image_content_type.to_sym
+
+  media_file_marketing_image_op = client.operation(:MediaFile)
+  media_file_marketing_image_op['create'] = media_file_marketing
+
+  response = media_file_service.mutate_media_files(
+    customer_id,
+    [
+      media_file_logo_op,
+      media_file_marketing_image_op,
+    ]
+  )
+
+  logo_id, marketing_image_id = response.results.to_a.map(&:resource_name)
+
+  gmail_ad = client.resource(:GmailAdInfo)
+  gmail_ad.teaser = client.resource(:GmailTeaser)
+  gmail_ad.teaser.headline = client.wrapper.string("Dream")
+  gmail_ad.teaser.description = client.wrapper.string(
+    "Create your own adventure"
+  )
+  gmail_ad.teaser.business_name = client.wrapper.string("Interplanetary Ships")
+  gmail_ad.teaser.logo_image = client.wrapper.string(logo_id)
+
+  gmail_ad.marketing_image = client.wrapper.string(marketing_image_id)
+  gmail_ad.marketing_image_headline = client.wrapper.string("Travel")
+  gmail_ad.marketing_image_description = client.wrapper.string("Take to the skies!")
+
+  ad = client.resource(:Ad)
+  ad.final_urls << client.wrapper.string("http://www.example.com")
+  ad.gmail_ad = gmail_ad
+  ad.name = client.wrapper.string("gmail ad")
+
+  ad_group_ad = client.resource(:AdGroupAd)
+  ad_group_ad.ad = ad
+  ad_group_ad.status = :PAUSED
+  ad_group_ad.ad_group = client.wrapper.string(
+    client.path.ad_group(customer_id, ad_group_id),
+  )
+
+  op = client.operation(:AdGroupAd)
+  op['create'] = ad_group_ad
+
+  ad_group_ad_service = client.service(:AdGroupAd)
+  response = ad_group_ad_service.mutate_ad_group_ads(customer_id, [op])
+  puts "created gmail ad with id #{response.results.first.resource_name}"
+end
+
+def get_image(url)
+  open(url) { |f|
```

Comment:
Aren't do/end preferred on multi-line blocks?

Reply:
I'm following the "semantic blocks" convention. This is an infinite yak shave in the ruby community tho. https://github.com/testdouble/standard/blob/master/README.md#standardrb-the-rules